### PR TITLE
Support model authorization in controllers.

### PR DESCRIPTION
**Pull request by robertpustulka**

### CODE REVIEW

File: tests/TestCase/Controller/Component/AuthorizationComponentTest.php [hunk 1]
```diff
@@ -104,4 +103,50 @@ public function testAuthorizeCustomException()
...
+
+    public function testAuthorizeModelSuccess()
+    {
+        $service = new AuthorizationService(new OrmResolver());
+        $identity = new IdentityDecorator($service, ['can_edit' => true]);
+        $this->Controller->request = $this->Controller->request
+            ->withAttribute('identity', $identity);
+
+        $this->Auth->authorizeModel();
+        $this->assertTrue(true);
```
Comment (markstory) [inline tests/TestCase/Controller/Component/AuthorizationComponentTest.php]: This doesn't look right.

File: src/Controller/Component/AuthorizationComponent.php [hunk 2]
```diff
@@ -86,4 +88,30 @@ protected function getIdentity(ServerRequestInterface $request)
...
     }
+
+    /**
+     * Model authorization handler.
+     *
+     * @return void
+     */
+    public function authorizeModel()
+    {
+        $this->authorize($this->getController()->loadModel());
```
Comment (markstory) [inline src/Controller/Component/AuthorizationComponent.php]: For a view/edit/delete operation does is it useful to check if the user has access in general? Personally I would like to know if the current user is allowed to modify/access the specific records they are working on.

File: tests/TestCase/Controller/Component/AuthorizationComponentTest.php (hunk 1 above)
Comment (robertpustulka) [reply tests/TestCase/Controller/Component/AuthorizationComponentTest.php]: I’m not sure how to assert that exception has not been thrown.

File: src/Controller/Component/AuthorizationComponent.php (hunk 2 above)
Comment (robertpustulka) [reply src/Controller/Component/AuthorizationComponent.php]: The access for individual entities should be checked using entity policies IMO.

So `$this->Authorization->authorize($entity)` in action body. I’m not a fan of loading entities magically and checking access. Feels like a job for a plugin, Crud for instance.

Comment (robertpustulka) [reply src/Controller/Component/AuthorizationComponent.php]: I can add a config option for supported actions if that’s your concern.

File: tests/TestCase/Controller/Component/AuthorizationComponentTest.php (hunk 1 above)
Comment (markstory) [reply tests/TestCase/Controller/Component/AuthorizationComponentTest.php]: You could assert the return of `authorizeModel`

File: src/Controller/Component/AuthorizationComponent.php (hunk 2 above)
Comment (markstory) [reply src/Controller/Component/AuthorizationComponent.php]: Yeah I don't want to magically load entities either. I could see this being useful for add at least.

File: tests/TestCase/Controller/Component/AuthorizationComponentTest.php (hunk 1 above)
Comment (jeremyharris) [reply tests/TestCase/Controller/Component/AuthorizationComponentTest.php]: What about `$this->assertNull($this->Auth->authorizeModel());`? Adds to assertion count, allows further assertions, and will fail if an exception is thrown.

**Comment by robertpustulka** [comment]
I've added `actionMap` config option. It allows to map controller action names to authorization actions as well as disabling authorization checks on certain actions. By default the controller action name is used.

```php
$this->loadComponent('Authorization, ['actionMap' => [
    'edit' => 'modify' //will use `canModify` auth check
    'index' => false //will not check access for `index` action
]];
```

**Pull request merged**